### PR TITLE
Customizable dependency failure handling for AsyncLoader

### PR DESCRIPTION
**Pull request by serxa**
Allows jobs to avoid failure in case of dependency failure. For example, database startup should not be canceled if one of the tables in the database is unable to load itself. It fixes a rare bug in server async startup.

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by serxa** [comment]
Look like failed stateless test were recently broken in master: https://s3.amazonaws.com/clickhouse-test-reports/0/d9a6096b323d69b3ce7197d20bd2a31aafbdb13a/stateless_tests__release__s3_storage__[2_2].html
(Hope they were not broken by previous work on async loading)

**Comment by serxa** [comment]
Looks like randomized unit tests have found some issues:
```
[ RUN      ] AsyncLoader.RandomIndependentTasks
Logical error: 'scheduled_jobs.contains(dpt)'.

Thread 1 "unit_tests_dbms" received signal SIGABRT, Aborted.
0x00007fc5d6c149fc in pthread_kill () from /lib/x86_64-linux-gnu/libc.so.6
#0  0x00007fc5d6c149fc in pthread_kill () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007fc5d6bc0476 in raise () from /lib/x86_64-linux-gnu/libc.so.6
#2  0x00007fc5d6ba67f3 in abort () from /lib/x86_64-linux-gnu/libc.so.6
#3  0x000056142a047108 in DB::abortOnFailedAssertion (description=...)
    at ./build_docker/./src/Common/Exception.cpp:44
#4  0x0000561429feb612 in DB::AsyncLoader::finish (
    this=this@entry=0x7ffd54fe9be0, job=..., 
    status=status@entry=DB::LoadStatus::CANCELED, reason=..., lock=...)
    at ./build_docker/./src/Common/AsyncLoader.cpp:572
#5  0x0000561429fe8248 in DB::AsyncLoader::remove (this=0x7ffd54fe9be0, 
    jobs=...) at ./build_docker/./src/Common/AsyncLoader.cpp:417
#6  0x0000561429fe7d23 in DB::LoadTask::remove (this=0x561439831608)
    at ./build_docker/./src/Common/AsyncLoader.cpp:171
#7  0x0000561429fe7c6e in DB::LoadTask::~LoadTask (this=0x13)
    at ./build_docker/./src/Common/AsyncLoader.cpp:152
#8  0x000056140d94eae0 in std::__1::__shared_count::__release_shared[abi:v15000]() (this=this@entry=0x5614398315f0)
    at ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:174
#9  0x000056140d94e9ef in std::__1::__shared_weak_count::__release_shared[abi:v15000]() (this=0x5614398315f0)
    at ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:215
#10 0x000056140db26dc8 in std::__1::shared_ptr<DB::LoadTask>::~shared_ptr[abi:v15000]() (this=0x561438644640)
    at ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:702
#11 std::__1::__destroy_at[abi:v15000]<std::__1::shared_ptr<DB::LoadTask>, 0>(std::__1::shared_ptr<DB::LoadTask>*) (__loc=0x561438644640)
    at ./contrib/llvm-project/libcxx/include/__memory/construct_at.h:63
#12 std::__1::destroy_at[abi:v15000]<std::__1::shared_ptr<DB::LoadTask>, 0>(std::__1::shared_ptr<DB::LoadTask>*) (__loc=0x561438644640)
    at ./contrib/llvm-project/libcxx/include/__memory/construct_at.h:88
#13 std::__1::allocator_traits<std::__1::allocator<std::__1::shared_ptr<DB::LoadTask> > >::destroy[abi:v15000]<std::__1::shared_ptr<DB::LoadTask>, void, void>(std::__1::allocator<std::__1::shared_ptr<DB::LoadTask> >&, std::__1::shared_ptr<DB::LoadTask>*) (__p=0x561438644640)
    at ./contrib/llvm-project/libcxx/include/__memory/allocator_traits.h:317
#14 std::__1::vector<std::__1::shared_ptr<DB::LoadTask>, std::__1::allocator<std::__1::shared_ptr<DB::LoadTask> > >::__base_destruct_at_end[abi:v15000](std::__1::shared_ptr<DB::LoadTask>*) (this=this@entry=0x7ffd54fe9de0, 
    __new_last=0x561438642650)
    at ./contrib/llvm-project/libcxx/include/vector:833
#15 0x000056140db11fe0 in std::__1::vector<std::__1::shared_ptr<DB::LoadTask>, std::__1::allocator<std::__1::shared_ptr<DB::LoadTask> > >::__clear[abi:v15000]() (this=0x7ffd54fe9de0) at ./contrib/llvm-project/libcxx/include/vector:827
#16 std::__1::vector<std::__1::shared_ptr<DB::LoadTask>, std::__1::allocator<std::__1::shared_ptr<DB::LoadTask> > >::~vector[abi:v15000]() (
    this=0x7ffd54fe9de0) at ./contrib/llvm-project/libcxx/include/vector:436
#17 AsyncLoader_RandomIndependentTasks_Test::TestBody (this=<optimized out>)
    at ./build_docker/./src/Common/tests/gtest_async_loader.cpp:909
#18 0x000056141a46934b in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void> (method=<optimized out>, 
    location=0x5614068f02ad "the test body", object=<optimized out>)
    at ./contrib/googletest/googletest/src/gtest.cc:2612
#19 testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void> (object=object@entry=0x561438b1de40, method=<optimized out>, 
    location=0x5614068f02ad "the test body")
    at ./contrib/googletest/googletest/src/gtest.cc:2648
#20 0x000056141a4692a1 in testing::Test::Run (this=this@entry=0x561438b1de40)
    at ./contrib/googletest/googletest/src/gtest.cc:2687
#21 0x000056141a46a6e0 in testing::TestInfo::Run (this=0x561437262910)
    at ./contrib/googletest/googletest/src/gtest.cc:2836
#22 0x000056141a46b795 in testing::TestSuite::Run (this=0x561437272530)
    at ./contrib/googletest/googletest/src/gtest.cc:3015
#23 0x000056141a47c9be in testing::internal::UnitTestImpl::RunAllTests (
    this=0x56143726d530) at ./contrib/googletest/googletest/src/gtest.cc:5921
#24 0x000056141a47c20b in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (method=<optimized out>, 
    location=0x5614067f0c42 "auxiliary test code (environments or event listeners)", object=<optimized out>)
    at ./contrib/googletest/googletest/src/gtest.cc:2612
#25 testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=<optimized out>, method=<optimized out>, 
    location=0x5614067f0c42 "auxiliary test code (environments or event listeners)") at ./contrib/googletest/googletest/src/gtest.cc:2648
#26 0x000056141a47c087 in testing::UnitTest::Run (
    this=0x5614352c2cc8 <testing::UnitTest::GetInstance()::instance>)
    at ./contrib/googletest/googletest/src/gtest.cc:5485
#27 0x000056140dc08d0f in RUN_ALL_TESTS ()
    at ./contrib/googletest/googletest/include/gtest/gtest.h:2316
#28 main (argc=1, argv=<optimized out>)
    at ./build_docker/./src/Common/tests/gtest_main.cpp:17
```

**Comment by serxa** [comment]
Here is another code path leading to the same problem. 
```
[ RUN      ] AsyncLoader.RandomDependentTasks
Logical error: 'scheduled_jobs.contains(dpt)'.

Thread 1 "unit_tests_dbms" received signal SIGABRT, Aborted.
0x00007f672ee6d9fc in pthread_kill () from /lib/x86_64-linux-gnu/libc.so.6
#0  0x00007f672ee6d9fc in pthread_kill () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007f672ee19476 in raise () from /lib/x86_64-linux-gnu/libc.so.6
#2  0x00007f672edff7f3 in abort () from /lib/x86_64-linux-gnu/libc.so.6
#3  0x0000559672d585a8 in DB::abortOnFailedAssertion (description=...)
    at ./build_docker/./src/Common/Exception.cpp:43
#4  0x0000559672cfb5f2 in DB::AsyncLoader::finish (
    this=this@entry=0x7ffeaa95f0a0, job=..., 
    status=status@entry=DB::LoadStatus::CANCELED, reason=..., lock=...)
    at ./build_docker/./src/Common/AsyncLoader.cpp:572
#5  0x0000559672cf9a2b in DB::AsyncLoader::schedule (this=0x7ffeaa95f0a0, 
    jobs_to_schedule=...) at ./build_docker/./src/Common/AsyncLoader.cpp:357
#6  0x00005596566ccbfd in AsyncLoaderTest::schedule (
    this=this@entry=0x7ffeaa95f0a0, jobs=...)
    at ./build_docker/./src/Common/tests/gtest_async_loader.cpp:147
#7  0x00005596566c1687 in AsyncLoader_RandomDependentTasks_Test::TestBody (
    this=<optimized out>)
    at ./build_docker/./src/Common/tests/gtest_async_loader.cpp:940
#8  0x00005596630ab8eb in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void> (method=<optimized out>, 
    location=0x55964f46a0ff "the test body", object=<optimized out>)
    at ./contrib/googletest/googletest/src/gtest.cc:2612
#9  testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void> (object=object@entry=0x559680e4fc20, method=<optimized out>, 
    location=0x55964f46a0ff "the test body")
    at ./contrib/googletest/googletest/src/gtest.cc:2648
#10 0x00005596630ab841 in testing::Test::Run (this=this@entry=0x559680e4fc20)
    at ./contrib/googletest/googletest/src/gtest.cc:2687
#11 0x00005596630acc80 in testing::TestInfo::Run (this=0x55967ea54300)
    at ./contrib/googletest/googletest/src/gtest.cc:2836
#12 0x00005596630add35 in testing::TestSuite::Run (this=0x55967ea51530)
    at ./contrib/googletest/googletest/src/gtest.cc:3015
#13 0x00005596630bef5e in testing::internal::UnitTestImpl::RunAllTests (
    this=0x55967ea4c530) at ./contrib/googletest/googletest/src/gtest.cc:5921
#14 0x00005596630be7ab in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (method=<optimized out>, 
    location=0x55964f36945c "auxiliary test code (environments or event listeners)", object=<optimized out>)
    at ./contrib/googletest/googletest/src/gtest.cc:2612
#15 testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=<optimized out>, method=<optimized out>, 
    location=0x55964f36945c "auxiliary test code (environments or event listeners)") at ./contrib/googletest/googletest/src/gtest.cc:2648
#16 0x00005596630be627 in testing::UnitTest::Run (
    this=0x55967e017220 <testing::UnitTest::GetInstance()::instance>)
    at ./contrib/googletest/googletest/src/gtest.cc:5485
#17 0x00005596567b228f in RUN_ALL_TESTS ()
    at ./contrib/googletest/googletest/include/gtest/gtest.h:2316
#18 main (argc=1, argv=<optimized out>)
    at ./build_docker/./src/Common/tests/gtest_main.cpp:17
```

**Comment by serxa** [comment]
I reintroduced the bug fixed here (at least for the first failure) https://github.com/ClickHouse/ClickHouse/commit/911c46aa3298d7ee32fd745e56261924ec67cf11

File: src/Common/AsyncLoader.h [hunk 1]
```diff
@@ -1,13 +1,15 @@
...
 #include <condition_variable>
+#include <concepts>
 #include <exception>
 #include <memory>
 #include <map>
 #include <mutex>
 #include <vector>
 #include <unordered_set>
 #include <unordered_map>
+#include <__concepts/invocable.h>
```
Comment (serxa) [inline src/Common/AsyncLoader.h:12]:
```suggestion
```

File: src/Common/AsyncLoader.cpp [hunk 2]
```diff
@@ -1,5 +1,6 @@
 #include <Common/AsyncLoader.h>
 
+#include <exception>
```
Comment (serxa) [inline src/Common/AsyncLoader.cpp:3]:
```suggestion
```

File: src/Databases/DatabaseOrdinary.cpp [hunk 3]
```diff
@@ -20,6 +20,7 @@
 #include <Parsers/formatAST.h>
 #include <Parsers/parseQuery.h>
 #include <Parsers/queryToString.h>
+#include "Common/AsyncLoader.h"
```
Comment (serxa) [inline src/Databases/DatabaseOrdinary.cpp:23]:
```suggestion
```

**Comment by serxa** [comment]
Data race seems unrelated at the first glance https://s3.amazonaws.com/clickhouse-test-reports/57697/e2b43c2cfadf552f173f0103ac28101d3aed6a87/stress_test__tsan_/stderr.log

File: src/Common/AsyncLoader.cpp [hunk 4]
```diff
@@ -612,6 +637,9 @@ void AsyncLoader::prioritize(const LoadJobPtr & job, size_t new_pool_id, std::un
     }
 
     job->pool_id.store(new_pool_id);
+    // TODO(serxa): we should adjust suspended_workers and suspended_waiters here.
+    // Otherwise suspended_workers we be left inconsistent. Fix it and add a test.
+    // Scenario: schedule a job A, wait for it from a job B in the same pool, prioritize A
```
Comment (tavplubix) [inline src/Common/AsyncLoader.cpp:642]: Is it for a separate PR?

**Review by tavplubix — APPROVED** [review]

**Comment by tavplubix** [comment]
Stress test (tsan) - https://github.com/ClickHouse/ClickHouse/issues/58437
Upgrade check (debug) - `failed at position 72 ('REFRESH')` - probably that because refreshable MVs were reverted just after the release, they are already merged back

File: src/Common/AsyncLoader.cpp (hunk 4 above)
Comment (serxa) [reply src/Common/AsyncLoader.cpp:642]: Yes https://github.com/ClickHouse/ClickHouse/pull/58362

**Pull request merged**